Mathlib/Algebra/Order/Ring/Ordering/Defs.lean

```diff
@@ -0,0 +1,197 @@
+but not containing `-1`. -/
+@[ext]
+structure RingPreordering extends Subsemiring R where
+  isSquare_mem' {x : R} (hx : IsSquare x) : x ∈ carrier
+  minus_one_not_mem' : -1 ∉ carrier
+
+namespace RingPreordering
+
+attribute [coe] toSubsemiring
+
+instance : SetLike (RingPreordering R) R where
+  coe P := P.carrier
+  coe_injective' p q h := by cases p; cases q; congr; exact SetLike.ext' h
+
+instance : SubsemiringClass (RingPreordering R) R where
+  zero_mem _ := Subsemiring.zero_mem _
+  one_mem _ := Subsemiring.one_mem _
+  add_mem := Subsemiring.add_mem _
+  mul_mem := Subsemiring.mul_mem _
+
+variable {R}
+
+@[aesop unsafe 50% (rule_sets := [SetLike])]
+protected theorem isSquare_mem (P : RingPreordering R) {x : R} (hx : IsSquare x) : x ∈ P :=
+  RingPreordering.isSquare_mem' _ hx
+
+@[simp, aesop safe (rule_sets := [SetLike])]
+protected theorem mul_self_mem (P : RingPreordering R) (x : R) : x * x ∈ P := by aesop
+
+@[simp, aesop safe (rule_sets := [SetLike])]
+protected theorem pow_two_mem (P : RingPreordering R) (x : R) : x ^ 2 ∈ P := by aesop
+
+@[aesop unsafe 20% forward (rule_sets := [SetLike])]
+protected theorem minus_one_not_mem (P : RingPreordering R) : -1 ∉ P :=
+  RingPreordering.minus_one_not_mem' _
+
+theorem toSubsemiring_injective :
+    Function.Injective (toSubsemiring : RingPreordering R → _) := fun A B h => by ext; rw [h]
+
+@[simp]
+theorem toSubsemiring_eq {P₁ P₂ : RingPreordering R} :
+    P₁.toSubsemiring = P₂.toSubsemiring ↔ P₁ = P₂ := toSubsemiring_injective.eq_iff
+
+@[simp]
+theorem mem_toSubsemiring {P : RingPreordering R} {x : R} :
+  x ∈ P.toSubsemiring ↔ x ∈ P := .rfl
+
+@[simp]
+theorem coe_toSubsemiring {P : RingPreordering R} :
+  (P.toSubsemiring : Set R) = P := rfl
+
+@[simp]
+theorem mem_mk {toSubsemiring : Subsemiring R}
+    (isSquare_mem) (minus_one_not_mem) {x : R} :
+    x ∈ mk toSubsemiring isSquare_mem minus_one_not_mem ↔ x ∈ toSubsemiring := .rfl
+
+@[simp]
+theorem coe_set_mk {toSubsemiring : Subsemiring R}
+    (isSquare_mem) (minus_one_not_mem) :
+    (mk toSubsemiring isSquare_mem minus_one_not_mem : Set R) = toSubsemiring := rfl
+
+section copy
+
+variable (P : RingPreordering R) (S : Set R) (hS : S = P)
+
+/-- Copy of a preordering with a new `carrier` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def copy : RingPreordering R where
+  carrier := S
+  zero_mem' := by aesop
+  add_mem' ha hb := by aesop
+  one_mem' := by aesop
+  mul_mem' ha hb := by aesop
+  isSquare_mem' := by aesop
+  minus_one_not_mem' := by aesop
+
+@[simp]
+theorem coe_copy : (P.copy S hS : Set R) = S := rfl
+
+theorem copy_eq : P.copy S hS = S := rfl
+
+end copy
+
+variable {P : RingPreordering R}
+
+/-!
+## Support
+-/
+
+namespace AddSubgroup
+
+variable (P) in
+/--
+The support of a ring preordering `P` in a commutative ring `R` is
+the set of elements `x` in `R` such that both `x` and `-x` lie in `P`.
+-/
+def support : AddSubgroup R where
```

Comment:
Why is this in the `AddSubgroup` namespace? I think `RingPreordering.support` is a much better name

Reply:
Aha, I see. In that case, we usually suffix the type to the non'main variants. Here that would give `RingPreordering.supportAddSubgroup : AddSubgroup _`, `RingPreordering.support : Ideal _`